Implement Deserialize/Serialize for Option

djc opened the pull request:
It'd be nice if I could handle `Option<&'a [u8]>` directly through `#[serde(with = "serde_bytes")]`.

dtolnay commented on src/ser.rs:
```diff
@@ -105,3 +105,41 @@ where
+        S: Serializer,
+    {
+        match self {
+            Some(b) => serializer.serialize_some(b),
```
This would serialize as a slice, not bytes.

dtolnay commented on src/de.rs:
```diff
@@ -103,3 +103,33 @@ impl<'de> Deserialize<'de> for Box<Bytes> {
+    where
+        D: Deserializer<'de>,
+    {
+        Deserialize::deserialize(deserializer).map(|v| Some(ByteBuf::into_vec(v)))
```
It doesn't look like this supports deserializing None.

djc commented:
I've simplified the impls to just wrap `Option<T>` around `T: Deserialize`/`T: Serialize`, and added a basic test for `Some` and `None`. Since this otherwise relies on the existing impls, it didn't seem like any more exhausting testing was needed, but let me know if you feel otherwise.

djc commented:
@dtolnay friendly reminder?

dtolnay commented on src/de.rs:
```diff
@@ -103,3 +103,15 @@ impl<'de> Deserialize<'de> for Box<Bytes> {
     }
 }
+
+impl<'de: 'a, 'a, T> Deserialize<'de> for Option<T>
```
The `'a` on this line is not used in the impl and can be removed.

djc commented:
Ah, you're right -- it's more subtle than I expected. I think this version solves the issues.

dtolnay commented on src/de.rs:
```diff
@@ -103,3 +103,16 @@ impl<'de> Deserialize<'de> for Box<Bytes> {
+
+impl<'de, T> Deserialize<'de> for Option<T>
+where
+    T: Deserialize<'de> + serde::de::Deserialize<'de> + From<&'de [u8]>,
```
Ideally this would only require `T: Deserialize<'de>`, not the other two bounds.

dtolnay commented on src/ser.rs:
```diff
@@ -105,3 +105,15 @@ where
     }
 }
+
+impl<T> Serialize for Option<T> where T: Serialize + std::ops::Deref<Target = [u8]> {
```
Ideally this would only require `T: Serialize`.

dtolnay commented on src/de.rs:
```diff
@@ -103,3 +103,16 @@ impl<'de> Deserialize<'de> for Box<Bytes> {
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes: Option<&[u8]> = serde::Deserialize::deserialize(deserializer)?;
```
This only works for deserializers that support deserializing &[u8], not for deserializers that support only owned bytes. For example it wouldn't work for Bincode when deserializing using [`bincode::deserialize_from`](https://docs.rs/bincode/1.2.1/bincode/fn.deserialize_from.html) because there isn't a way to borrow bytes from a std::io::Read impl.

dtolnay commented:
I merged an alternative implementation in #21 and published 0.11.4 with the impls. Thanks!

djc commented:
Thanks for finishing this up, and sorry for being a bit iterative and not doing more upfront testing.
